src/rust/aggregator/src/pipelines/spreads.rs

```diff
@@ -0,0 +1,373 @@
+                "sqlx_queries/spreads/get_change_order_size_events_between_txn_versions.sql",
+                txn_version_of_state,
+                txn_version
+            )
+            .fetch_all(transaction as &mut PgConnection)
+            .await
+            .map_err(to_pipeline_error)?;
+
+            // Get all cancel events that happened between the last transaction included in the previous
+            // state and the last transaction that happened before the given timestamp.
+            let cancels = sqlx::query_file!(
+                "sqlx_queries/spreads/get_cancel_order_events_between_txn_versions.sql",
+                txn_version_of_state,
+                txn_version
+            )
+            .fetch_all(transaction as &mut PgConnection)
+            .await
+            .map_err(to_pipeline_error)?;
+
+            // Insert all places as orders into the state.
+            for place in places {
+                orders.insert((place.market_id.clone(), place.order_id.clone()), place);
+            }
+
+            // Remove all orders that have been cancelled.
+            for cancel in cancels {
+                orders.remove(&(cancel.market_id, cancel.order_id));
+            }
+
+            // Handle fills and changes chronologically.
+            let mut fills_index = 0;
+            let mut changes_index = 0;
+            for _ in 0..(fills.len() + changes.len()) {
+                let fill = fills.get(fills_index);
+                let change = changes.get(changes_index);
+                let (f, c) = match (fill, change) {
+                    (Some(fill), Some(change)) => {
+                        if (fill.txn_version.clone(), fill.event_idx.clone())
+                            < (change.txn_version.clone(), change.event_idx.clone())
+                        {
+                            (Some(fill), None)
+                        } else {
+                            (None, Some(change))
+                        }
+                    }
+                    (None, None) => unreachable!(),
+                    other => other,
+                };
+                match (f, c) {
+                    (Some(fill), None) => {
+                        handle_fill(fill, &mut orders, &mut fills_index);
+                    }
+                    (None, Some(change)) => {
+                        handle_change(change, &mut orders, &mut changes_index);
+                    }
+                    _ => unreachable!(),
+                };
+            }
+
+            txn_version_of_state = txn_version;
+        }
+    }
+    fn get_spreads_form_state<'a>(
+        &self,
+        orders: &HashMap<OrderKey, Order>,
+    ) -> (HashMap<BigDecimal, BigDecimal>, HashMap<BigDecimal, BigDecimal>) {
+
+        let mut min_asks: HashMap<BigDecimal, BigDecimal> = Default::default();
+        let mut max_bids: HashMap<BigDecimal, BigDecimal> = Default::default();
+
+        for element in orders.values().cloned() {
```

Comment:
Avoid cloning if possible, there's a decent amount of state here

Reply:
There is some cloning that needs to be done anyways.

It's either that, or:

```rust
for element in state.orders.values() {
    if element.side {
        if let Some(min) = min_asks.get(&element.market_id) {
            if *min > element.price {
                min_asks.insert(element.market_id.clone(), element.price.clone());
            }
        } else {
            min_asks.insert(element.market_id.clone(), element.price.clone());
        }
    } else {
        if let Some(max) = max_bids.get(&element.market_id) {
            if *max < element.price {
                max_bids.insert(element.market_id.clone(), element.price.clone());
            }
        } else {
            max_bids.insert(element.market_id.clone(), element.price.clone());
        }
    }
}
```

Even though in the second version, we only clone `price` and `market_id` instead of the entire order struct, I'm wondering if due to the way cloning works, it isn't more optimized to clone the whole array at once.

Reply:
Cloning `price` and `market_id` only is manageable, and the clone only happens when a new min/max is found so only some of the data will get cloned. This new approach seems more efficient